core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -866,6 +852,44 @@ class Partition(val topicPartition: TopicPartition,
     }
   }
 
+  def maybeLogSlowReplication(curMaximalIsr: Set[Int], requiredOffset: Long, requestCreationTime: Long, topicPartition: TopicPartition): Unit = {
+    val now = time.milliseconds()
+    val timeElapsedMs = time.milliseconds() - requestCreationTime
+
+    if (timeElapsedMs < logSlowReplicationThresholdMs) {
+      // Only slow replications is worth logging
+      return
+    }
+
+    // Check and maybe create new time bucket for log slow replication
+    if (now - logSlowReplicationBucketStartTime > logSlowReplicationBucketLengthMs) {
+      logSlowReplicationBucketCounter = 0
+      logSlowReplicationBucketStartTime = now
+    }
+
+    if (logSlowReplicationBucketCounter > logSlowReplicationBucketMaxLogCount) {
+      // Stop logging if there are already many logs in a time bucket. This is to avoid logs overwhelming the host.
+      return
+    }
+
+    def logEndOffsetString: ((Int, Long)) => String = {
+      case (brokerId, logEndOffset) => s"broker $brokerId: $logEndOffset"
+    }
+
+    val curInSyncReplicaObjects = (curMaximalIsr - localBrokerId).flatMap(getReplica)
+    val replicaInfo = curInSyncReplicaObjects.map(replica => (replica.brokerId, replica.logEndOffset))
+    val localLogInfo = (localBrokerId, localLogOrException.logEndOffset)
+    val (ackedReplicas, awaitingReplicas) = (replicaInfo + localLogInfo).partition {
+      _._2 >= requiredOffset
+    }
+
+    info(s"Progress awaiting ISR acks for partition ${topicPartition} offset $requiredOffset: " +
+      s"acked: ${ackedReplicas.map(logEndOffsetString)}, " +
+      s"awaiting ${awaitingReplicas.map(logEndOffsetString)}. " +
+      s"Already waited for ${timeElapsedMs} milliseconds so far")
+    logSlowReplicationBucketCounter += 1
+  }
```

Comment:
What about the correlation ID and/or other client-request info? Is that available here? At least the former would be very useful in specific cases (e.g., Espresso complaint about a specific slow request), but I don't know offhand whether it's available down here.

Reply:
Doubled checked and looks like those information are not available at this point. I'll skip them for now.